src/re/multi_grid/indexing.py

```diff
@@ -268,13 +268,37 @@ def _stack_outer(*arrays, outer_axis=-1, stack_axis=0):
     return res
 
 
+def regularAxis_index2cart(index, size):
+    # Map onto bincenter
+    cc = (index + 0.5) / size
+    # Add dummy axis for 1D space
+    return cc[np.newaxis]
+
+
+def healpixAxis_index2cart(index, nside, nest):
+    from healpy.pixelfunc import pix2vec
+    shp = index.shape[1:]
+    cc = pix2vec(nside, np.ravel(index), nest=nest)
+    return np.stack(cc, axis=0).reshape((3,) + shp)
+
+
+def default_index2cart(index, grids):
+    coords = tuple(
+        healpixAxis_index2cart(index[i], g.nside, g.nest)
+        if isinstance(g, HEALPixAxisAtLevel) else
+        regularAxis_index2cart(index[i], g.size) for i, g in enumerate(grids)
+    )
+    return coords #TODO better concatenate instead of tuple?
+
+
 @dataclass(kw_only=True)
 class OGridAtLevel:
     grids: tuple[GridAxisAtLevel]
     index2cart: callable
 
-    def __init__(self, *grids):
+    def __init__(self, *grids, index2cart=default_index2cart):
         self.grids = tuple(grids)
+        self.index2cart = index2cart
```

Comment:
note to self; think about whether grids have shared `index2cart` or every grid has its own `index2cart`.

At first glance, associating the "charting" (`index2cart` methods) with `GridAxis` seems like a good fit. `GridAxis` could be a base class and one needs to inherent from it to define a new charting. `HEALPixAxis` kind of automatically implies a natural Cartesian mapping but for all other `GridAxis` we pretend that it doesn't need a charting just yet even though neighbors are inherently coupled to the way the charting works.

Furthermore, I don't think there will be a lot of different axes. A regular, logarithmic and maybe some symmetric logarithmic chart will likely cover almost all use-case. It also forces the user to merge charts that couples indices and allows us to be more narrow in the treatment of all grids later on.

Reply:
You raise a very good point here! 
Let me add a few cents on what I encountered on the way:
- When implementing the "default" maps I initially thought about adding them as staticmethods to the `GridAxis`, it just seemed natural to put them there. Maybe having the concept of Axes as an entity and then some generic function that "maps indices to locs" is not the right way to go.
- We have to cover non-separable mappings in an easy to deal with way for the user. Can `OGrid` inherit from `GridAxis` in a way? Maybe a `BaseAxis` and structures that built on top of that does indeed make more sense after all.
- For your `HEALPixAxis` comment: Bare in mind that some might want to use just a single sphere with angular coordinates so we should not make this too difficult. This also raises the (somewhat related) question of volume and metric. I believe ultimately there is going to be a function for `intex2volume`, to which this discussion imho applies analogously.